### ISSUE TITLE
Going forward with CSP

### ISSUE THREAD

**Issue by rugk** [PrivateBin/PrivateBin#108]
Google has just released a [CSP testing tool](https://csp-evaluator.withgoogle.com/) and some more [detailed information about CSP](https://csp.withgoogle.com/docs/index.html).

They mention some interesting things:
1. There is a new ['strict-dynamic'](https://www.chromestatus.com/feature/5633814718054400) being created.
2. `unsafe-inline` can always be used as a fallback for older browsers supporting only CSPv2. CSPv3-compliant browsers will ignore this if `strict-dynamic`, nonces or hashes are used.
3. They do [discourage](https://csp.withgoogle.com/docs/strict-csp.html) the use of `'self'` in `script-src` as [it can be bypassed](https://csp.withgoogle.com/docs/faq.html#problems-with-whitelists) in certain cases.

In CSPv3 you can AFAIK also define subdirectories (`privatebin.org/scripts`), which could be another way to strengthen our policy. However in this case there might be compatibility issues.

So together with https://github.com/PrivateBin/PrivateBin/issues/82 this might be an issue we can address in the future when CSP has further developed itself :smiley:

**Comment by Framartin** [comment]
Agreed. Using `nonce-` or `sha256-` in `script-src` is a a lot better than `'self'` to mitigate the exploitation of XSS, like #183. See [Documentation](https://content-security-policy.com/#source_list).

**Comment by rugk** [comment]
Interesting, there is a new [`strict-dynamic`](https://content-security-policy.com/strict-dynamic/) thing in CSPv3 that – as far as I understand it – let's one script load any other scripts.
I'm still thinking of whether that could make it easier to use nounces or in our case, rather, hashes (we do recalculate SRI hashes anyway, so another place for these would also be possible).

**Comment by rugk** [comment]
And [there is more crazy (experimental) stuff](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/require-trusted-types-for), but that seems to look like a better solution for `unsafe-eval`, which we only include, because of Chrom/ium's PDF reader, so I guess we can ignore that…

**Comment by elrido** [comment]
The `unsafe-eval` is required in Chrome and Safari for WASM loading. If not set, calling WebAssembly.instantiate will trigger an error in those browsers.
https://github.com/PrivateBin/PrivateBin/blob/8232dce395e421450779bd710e5872399f792085/js/zlib-1.2.11.js#L34-L35

- [Technical background](https://github.com/WebAssembly/content-security-policy/blob/master/proposals/CSP.md#summary-of-webassembly-apis-and-their-risks)
- [Browser comparison matrix](https://github.com/WebAssembly/content-security-policy/blob/master/proposals/CSP.md#behavior-of-current-implementations)
- [Proposal to W3C to standardize `wasm-eval` directive](https://github.com/w3c/webappsec-csp/pull/293), that Chrome has implemented for web apps, but not web sites, that would allow to restrict eval while allowing wasm. It hasn't been accepted and even if it will, it will take years for browsers with support for it to be rolled out for a majority of users. :-(

**Comment by rugk** [comment]
Regarding [`unsafe-eval` the Matrix Element client has the same issue and someone suggested](https://github.com/vector-im/element-web/issues/12262#issuecomment-864274860) this:
> Using `WebAssembly.instantiateStreaming` would avoid this problem.

**Comment by rugk** [comment]
Also [the docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/instantiateStreaming) for this say:
> This is the most efficient, optimized way to load wasm code.

So would likely not be a bad idea to use this altogether…

**Comment by rugk** [comment]
Opened a new issue for that, so we can keep this “meta-issue” here and fix that WebAssembly loading: https://github.com/PrivateBin/PrivateBin/issues/814
